educational/systems/random_agent.py

```diff
@@ -0,0 +1,370 @@
+# limitations under the License.
+import logging
+import time
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Any, Dict, Optional, Tuple, Union
+
+import chex
+import jax
+import jax.numpy as jnp
+from absl import app, flags
+
+from mava import specs as mava_specs
+from mava.utils.environments import debugging_utils, smac_utils
+from mava.utils.loggers import logger_utils
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_string("system", "test agent", "What agent is running.")
+flags.DEFINE_string(
+    "base_dir", "~/mava", "Base dir to store experiment data e.g. checkpoints."
+)
+flags.DEFINE_string(
+    "env_name",
+    "simple_spread",
+    "Env name e.g. sinple_spread, 3m or 8m.",
+)
+flags.DEFINE_string(
+    "env_type",
+    "debug",
+    "Env type e.g. debug or smac.",
+)
+flags.DEFINE_string(
+    "env_action_type",
+    "discrete",
+    "Discrete or continous. Only applies for debug env,\
+    other env have their predefined types.",
+)
+flags.DEFINE_integer(
+    "max_total_steps",
+    100000,
+    "Max total steps (across all episodes).",
+)
+
+flags.DEFINE_integer(
+    "num_episodes",
+    None,
+    "Max number of episodes.",
+)
+
+
+@dataclass
+class InitConfig:
+    seed: int = 42
+
+
+@dataclass
+class EnvironmentConfig:
+    env_name: str
+    type: str
+    action_space: str
+    seed: int = 42
+
+
+@dataclass
+class SystemConfig:
+    max_total_steps: int
+    num_episodes: int
+    name: str = "random"
+    seed: int = 42
+
+
+@chex.dataclass(frozen=True, mappable_dataclass=False)
+class RandomSystemState:
+    rng: jnp.ndarray
+
+
+def init(config: InitConfig = InitConfig()) -> InitConfig:
+    """Init system.
+
+    This would handle thing to be done upon system once in the beginning of a run,
+    e.g. set random seeds.
+
+    Args:
+        config : init config.
+    """
+    return config
+
+
+def make_environment(
+    config: EnvironmentConfig,
+) -> Tuple[Any, EnvironmentConfig]:
+    """Init and return environment or wrapper.
+
+    Args:
+        config : env config.
+
+    Returns:
+        env or wrapper.
+    """
+    if config.type == "debug":
+        env, _ = debugging_utils.make_environment(
+            env_name=config.env_name,
+            action_space=config.action_space,
+            random_seed=config.seed,
+        )
+    elif config.type == "smac":
+        env, _ = smac_utils.make_environment(
+            map_name=config.env_name, random_seed=config.seed
+        )
+    return env, config
+
+
+def make_system(
+    environment_spec: mava_specs.MAEnvironmentSpec,
+    config: SystemConfig,
+) -> Tuple[Any, SystemConfig]:
+    """Inits and returns system/networks.
+
+    Args:
+        config : system config.
+
+    Returns:
+        system.
+    """
+    init_rng = jax.random.PRNGKey(config.seed)
+    agent_specs = environment_spec.get_agent_environment_specs()
+
+    @dataclass
+    class RandomExecutor:
```

Comment:
Ok after starting implementation I think this is too much boilerplate for single file systems, I think they should just be pretty messy to start with and once they are all implemented we can decide on a common interface.

Especially the observe and observe first, since a simple replay buffer will just have a single add, not necessarily an add first.

I think stand alone global methods, like `get_actions(obs, actor, params)` and `learn(batch, actor, actor_params, critic, ...)` would be a better approach, just to put as few constraints on design as possible and then we can refactor all implementations to a common interface later, once we've seen all implementations and can take the best parts from the ones we like.

What do you think?

Reply:
I agree with @sash-a too much boilerplate code. I think we really want to make these single file implementations as simple as possible.